feat(telemetry): report usage of custom registry certificates

vzhukovs opened the pull request:
### What does this PR do?

The purpose of these changes is to add a boolean flag to the telemetry initialization request indicating whether the user is using custom certificates.

### Screenshot / video of UI

N/A

### What issues does this PR fix or reference?

https://github.com/containers/podman-desktop-internal/issues/588

### How to test this PR?

Use the following steps to test this PR(macOS tested usecase):

**1. Create certificates on macOS**

```
mkdir -p ~/test-registry/certs
```

```
openssl req -x509 -newkey rsa:2048 -nodes -days 365 \
  -keyout ~/test-registry/certs/ca.key \
  -out   ~/test-registry/certs/ca.crt \
  -subj "/CN=Local Test CA"
```

```
openssl req -newkey rsa:2048 -nodes \
  -keyout ~/test-registry/certs/registry.key \
  -out   ~/test-registry/certs/registry.csr \
  -subj "/CN=registry.test"
```

```
cat > /tmp/registry-test.cnf <<'EOF'
[ req ]
distinguished_name = req_distinguished_name
x509_extensions    = v3_req
prompt = no

[ req_distinguished_name ]
CN = registry.test

[ v3_req ]
subjectAltName = @alt_names

[ alt_names ]
DNS.1 = registry.test
EOF
```

```
openssl x509 -req -in ~/test-registry/certs/registry.csr \
  -CA  ~/test-registry/certs/ca.crt \
  -CAkey ~/test-registry/certs/ca.key -CAcreateserial \
  -out ~/test-registry/certs/registry.crt -days 365 \
  -extfile /tmp/registry-test.cnf -extensions v3_req
```

**2. Add registry.test to /etc/hosts (both Mac and VM)**

```
echo '127.0.0.1 registry.test' | sudo tee -a /etc/hosts
echo '::1 registry.test'       | sudo tee -a /etc/hosts

podman machine ssh 'echo "127.0.0.1 registry.test" | sudo tee -a /etc/hosts'
```

**3. Prepare CA for Podman**

```
mkdir -p ~/.config/containers/certs.d/registry.test:5001
cp ~/test-registry/certs/ca.crt ~/.config/containers/certs.d/registry.test:5001/ca.crt

podman machine stop && podman machine start
```

**4. Copy certificates into VM**

```
podman machine ssh 'sudo mkdir -p /opt/registry/certs'

podman machine ssh 'cat | sudo tee /opt/registry/certs/registry.crt >/dev/null' < ~/test-registry/certs/registry.crt
podman machine ssh 'cat | sudo tee /opt/registry/certs/registry.key >/dev/null' < ~/test-registry/certs/registry.key

podman machine ssh 'sudo chmod 644 /opt/registry/certs/registry.crt'
podman machine ssh 'sudo chmod 600 /opt/registry/certs/registry.key'
```

**5. Create volume and transfer certificates into this volume**

```
podman volume rm -f reg-certs 2>/dev/null || true
podman volume create reg-certs
podman run --rm -i -v reg-certs:/certs alpine sh -c 'cat > /certs/registry.crt' \
  < <(podman machine ssh 'sudo cat /opt/registry/certs/registry.crt')

podman run --rm -i -v reg-certs:/certs alpine sh -c 'cat > /certs/registry.key' \
  < <(podman machine ssh 'sudo cat /opt/registry/certs/registry.key')

podman run --rm -v reg-certs:/certs alpine sh -c 'chmod 644 /certs/registry.crt && chmod 600 /certs/registry.key && ls -l /certs'
```

**6. Init local registry**

```
podman rm -f reg 2>/dev/null || true

podman run -d --name reg -p 5001:5000 \
  -v reg-certs:/certs:ro \
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/registry.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/registry.key \
  -e REGISTRY_LOG_LEVEL=debug \
  registry:2

podman ps -a
podman logs reg | tail -n +50
```

**7. Check registry with self-signed certificates works**

```
curl -vk https://registry.test:5001/v2/
podman login registry.test:5001 -u test -p test
# → Login Succeeded!
```

**8. Run Podman Desktop from the current branch**

**9. Obtain device ID**

```
cat ~/.redhat/anonymousId
```

**10. Navigate to Amplitude -> Users -> User Profiles and filter your device by Device ID from Step 9**

**11. Open your device and take a look to the events list**

**12. You should see new event called `podman-desktop.podman.custom_certs` which contains necessary information about using custom certificates and platform associated with.**

**13. Cleanup**

```
podman rm -f reg
podman volume rm -f reg-certs
podman image prune -f
podman container prune -f

podman machine ssh 'sudo rm -rf /opt/registry/certs'
rm -rf ~/.config/containers/certs.d/registry.test:5001
rm -rf ~/test-registry
```

**14. Restart Podman Desktop**

**15. Repeat Step 10-11**

**16. You should see that event `podman-desktop.podman.custom_certs` doesn't indicate that user is using custom certificates**

- [x] Tests are covering the bug fix or the new feature

eqqe commented on packages/main/src/plugin/telemetry/telemetry.spec.ts:
```diff
@@ -368,3 +376,60 @@ describe('aggregateTrack', () => {
+    },
+  ];
+
+  test.each(scenarios)('$name', async ({ setup, expected }) => {
```
It seems that this test is failing on windows, can anyone with this OS help to investigate?

benoitf commented on packages/main/src/plugin/telemetry/telemetry.spec.ts:
```diff
@@ -368,3 +376,60 @@ describe('aggregateTrack', () => {
+
+function safeSubdir(subdir?: string): string {
+  if (!subdir) return '';
+  return process.platform === 'win32' ? subdir.replace(/:/g, '_') : subdir;
```
we do have a util class with isWindows function

benoitf reviewed (dismissed):
it can't be a global flag

here you're reading a specific podman attribute (from .containers folder)

so it can't be a report of Podman Desktop, it needs to be reported by the podman extension

benoitf commented:
if I read the issue, it's about custom certificates but it does not seem to be related to have certificates for custom OCI registries

so I'm unsure this PR is matching the requested issue

vzhukovs commented:
@benoitf @Firewall I've updated PR, moved all logic to the Podman extension and now we can see a new event in aplitude. Also updated steps 11, 12 and 16 in PR description.

Firewall reviewed (approved):
Added a log in my console because I had issues sending my developer data to amplitude, but it is working! The instructions you gave worked flawlessly, thanks! 
<img width="786" height="204" alt="image" src="https://github.com/user-attachments/assets/b6473c44-7a58-48d5-b95b-287fd287859d" />

benoitf commented on extensions/podman/packages/extension/src/certificate-detection/certificate-detection-service.ts:
```diff
@@ -0,0 +1,243 @@
+      const result = await this.performDetection();
+      result.scanDurationMs = Date.now() - startTime;
+
+      if (this.config.enableTelemetry && this.telemetryLogger) {
+        await this.sendTelemetry(result);
+      }
```
I would move the compute in the if telemetry block

because why collect data if we don't need it ?

it'll simplify the logic as well in the catch as we check again if telemetry is enabled

benoitf commented on extensions/podman/packages/extension/src/certificate-detection/certificate-detection-service.ts:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  private async sendTelemetry(result: CertificateDetectionResult): Promise<void> {
+    if (!this.telemetryLogger) {
+      return;
+    }
```
I think we have redundant check at different place. either we check before calling or inside the method but both looks it little bit redundant

vzhukovs replied:
Yes, you're right, removed redundant check

benoitf commented on extensions/podman/packages/extension/src/extension.ts:
```diff
@@ -2278,3 +2292,24 @@ export function updateWSLHyperVEnabledContextValue(value: boolean): void {
+  // Initial detection
+  await certificateDetectionService.detectCustomCertificates();
+
+  // Set up periodic detection (every 24 hours)
```
do we need this check every 24hour or just once per launch ?

vzhukovs replied:
Theoretically, user can hold podman desktop opened for a long time and user can configure custom registry during this period, and it would be nice to handle the registry configuration changes.
